File: core/x/javaapi/api/src/main/java/viaduct/java/api/context/ResolverExecutionContext.java
package viaduct.java.api.context;

import java.util.Map;
import java.util.concurrent.CompletableFuture;
import viaduct.java.api.documents.MutationFromAnnotation;
import viaduct.java.api.documents.QueryFromAnnotation;
import viaduct.java.api.globalid.GlobalID;
import viaduct.java.api.reflect.Type;
import viaduct.java.api.types.GraphQLObject;
import viaduct.java.api.types.NodeCompositeOutput;
import viaduct.java.api.types.NodeObject;

/** A generic context for resolving fields or types. */
public interface ResolverExecutionContext extends ExecutionContext {
  /**
   * Creates a lazy reference to a Node from its GlobalID.
   *
   * <p>Only the {@code id} field is accessible on the returned reference; reading any other field
   * throws. Use it to build resolver responses for fields with Node types.
   *
   * <p>Example: {@code ctx.ref(ctx.globalIDFor(Type.ofClass(Planet.class), planetId))}.
   *
   * @param id the Node's GlobalID
   * @param <T> the Node GRT type
   * @return an unresolved reference GRT
   */
  <T extends NodeCompositeOutput> T ref(GlobalID<T> id);

  /**
   * Creates a lazy reference to an object-valued query root field or a field reachable through
   * {@code @namespaceType} fields.
   *
   * <p>The engine resolves the reference after the current resolver returns it directly or embeds
   * it in another GRT. Fields cannot be read from the unresolved reference.
   *
   * <p>Example: {@code ctx.ref(ProductFactory.create())}.
   *
   * @param call the generated root field call
   * @param <T> the field's object output type
   * @return an unresolved reference GRT
   */
  <T extends GraphQLObject> T ref(RootFieldCall<T> call);

  /**
   * Creates a serialized GlobalID string for the given Node type and internal ID.
   *
   * <p>Use this instead of {@link #globalIDFor} when you need a serialized GlobalID string (e.g.,
   * to set the {@code id} field on a Node type builder).
   *
   * <p>Example: {@code ctx.globalIDStringFor(Type.ofClass(NodeObj.class), internalId)}
   *
   * @param type the GraphQL Node type
   * @param internalID the internal ID string
   * @param <T> the Node GRT type
   * @return the serialized GlobalID string
   */
  <T extends NodeObject> String globalIDStringFor(Type<T> type, String internalID);

  /**
   * Executes a subquery against the Query root type and returns the result as an instance of {@code
   * targetClass} with the requested selections populated.
   *
   * @param selections the GraphQL selection string (e.g., {@code "firstName lastName"})
   * @param variables variables map for parameterized selections
   * @param targetClass the Java GRT class to populate with the resolved data
   * @param <T> the Java GRT type for the Query root
   * @return a future that completes with the populated Query GRT instance
   */
  <T> CompletableFuture<T> query(
      String selections, Map<String, Object> variables, Class<T> targetClass);

  /** Executes a subquery against the Query root type with no variables. */
  default <T> CompletableFuture<T> query(String selections, Class<T> targetClass) {
    return query(selections, Map.of(), targetClass);
  }

  /**
   * Executes a query declared with {@code @GraphQLOperation}.
   *
   * <p>Generated resolver contexts provide overloads that infer {@code targetClass} from the
   * tenant's Query GRT.
   */
  default <T> CompletableFuture<T> query(
      QueryFromAnnotation operation, Map<String, Object> variables, Class<T> targetClass) {
    return query(operation.getOperationText(), variables, targetClass);
  }

  /** Executes an annotated query with no variables. */
  default <T> CompletableFuture<T> query(QueryFromAnnotation operation, Class<T> targetClass) {
    return query(operation, Map.of(), targetClass);
  }

  /**
   * Executes a sub-mutation against the Mutation root type and returns the result as an instance of
   * {@code targetClass} with the requested selections populated.
   *
   * @param selections the GraphQL selection string (e.g., {@code "incrementCounter"})
   * @param variables variables map for parameterized selections
   * @param targetClass the Java GRT class to populate with the resolved data
   * @param <T> the Java GRT type for the Mutation root
   * @return a future that completes with the populated Mutation GRT instance
   */
  <T> CompletableFuture<T> mutation(
      String selections, Map<String, Object> variables, Class<T> targetClass);

  /** Executes a sub-mutation against the Mutation root type with no variables. */
  default <T> CompletableFuture<T> mutation(String selections, Class<T> targetClass) {
    return mutation(selections, Map.of(), targetClass);
  }

  /**
   * Executes a mutation declared with {@code @GraphQLOperation}.
   *
   * <p>Generated resolver contexts provide overloads that infer {@code targetClass} from the
   * tenant's Mutation GRT.
   */
  default <T> CompletableFuture<T> mutation(
      MutationFromAnnotation operation, Map<String, Object> variables, Class<T> targetClass) {
    return mutation(operation.getOperationText(), variables, targetClass);
  }

  /** Executes an annotated mutation with no variables. */
  default <T> CompletableFuture<T> mutation(
      MutationFromAnnotation operation, Class<T> targetClass) {
    return mutation(operation, Map.of(), targetClass);
  }
}

File: core/tenant/api/src/main/kotlin/viaduct/api/context/ResolverExecutionContext.kt
package viaduct.api.context

import viaduct.api.documents.QueryFromAnnotation
import viaduct.api.documents.Selections
import viaduct.api.globalid.GlobalID
import viaduct.api.reflect.Type
import viaduct.api.select.SelectionSet
import viaduct.api.types.CompositeOutput
import viaduct.api.types.NodeObject
import viaduct.api.types.Object
import viaduct.api.types.Query as QueryType
import viaduct.apiannotations.ExperimentalApi
import viaduct.apiannotations.StableApi

/** A generic context for resolving fields or types */
@StableApi
interface ResolverExecutionContext<Q : QueryType> : ExecutionContext {
    /**
     * Loads the operation declared by a [@GraphQLOperation][viaduct.api.documents.GraphQLOperation]
     * query object on the root Query type, and returns the response typed as [Q].
     *
     * Example usage:
     * ```
     * val result = ctx.query(GetUserQuery)
     * ```
     *
     * @param operation The query operation object declaring the operation document
     * @param variables Optional variables to use in the operation
     * @return The query result typed as [Q]
     */
    @StableApi
    suspend fun query(
        operation: QueryFromAnnotation,
        variables: Map<String, Any?> = emptyMap()
    ): Q

    /**
     * Creates a [SelectionSet] on a provided type from the provided [Selections] String
     * @see [Selections]
     */
    fun <T : CompositeOutput> selectionsFor(
        type: Type<T>,
        selections: @Selections String,
        variables: Map<String, Any?> = emptyMap()
    ): SelectionSet<T>

    /**
     * Creates a GlobalID and returns it as a String. Example usage:
     *   globalIDStringFor(User.Reflection, "123")
     */
    fun <T : NodeObject> globalIDStringFor(
        type: Type<T>,
        internalID: String
    ): String

    /**
     * Creates a lazy reference given a Node ID. Only the ID field is accessible from the
     * created reference. Attempting to access other fields will result in an exception.
     * This can be used to construct resolver responses for fields with Node types.
     */
    fun <T : NodeObject> ref(id: GlobalID<T>): T

    /**
     * Creates a lazy reference to a root field that will be later resolved by the engine.
     *
     * No fields are accessible from the returned value. Attempting to access fields will result in
     * an exception. The returned value may be used as a resolver return value or passed to a GRT
     * builder.
     *
     * Example usage:
     * ```
     * val product = ctx.ref(ProductFactory.create { name("Air") })
     * ```
     */
    @ExperimentalApi
    fun <T : Object> ref(call: RootFieldCall<T>): T
}
